File: experiments-local/experiment-17/runs/agent=omp_language=typescript_model=openrouter/deepseek/deepseek-v4-pro_tooling=none/rep1/src/queries.ts
/**
 * Brazilian Soccer MCP Server - Query Engine
 *
 * All data query functions: match search, team stats, player search,
 * competition standings, head-to-head, and statistical analysis.
 */

import { getMatches, getPlayers } from './data.js';
import { normalizeTeam } from './normalize.js';
import type {
  NormalizedMatch,
  NormalizedPlayer,
  MatchQuery,
  PlayerQuery,
  TeamRecord,
  TeamStats,
  HeadToHead,
  StandingEntry,
} from './types.js';

// ── Helpers ──────────────────────────────────────────────────────────

function teamMatchesQuery(
  team: NormalizedMatch,
  normalizedTeam: string
): boolean {
  return team.homeTeam === normalizedTeam || team.awayTeam === normalizedTeam;
}

function emptyTeamStats(team: string, display: string): TeamStats {
  return { team, teamDisplay: display, matches: 0, wins: 0, draws: 0, losses: 0, goalsFor: 0, goalsAgainst: 0, points: 0 };
}

function calcStats(teamName: string, display: string, matches: NormalizedMatch[]): TeamStats {
  const stats = emptyTeamStats(teamName, display);
  for (const m of matches) {
    const isHome = m.homeTeam === teamName;
    stats.matches++;
    stats.goalsFor += isHome ? m.homeGoal : m.awayGoal;
    stats.goalsAgainst += isHome ? m.awayGoal : m.homeGoal;
    if (m.homeGoal === m.awayGoal) {
      stats.draws++;
      stats.points += 1;
    } else if ((isHome && m.homeGoal > m.awayGoal) || (!isHome && m.awayGoal > m.homeGoal)) {
      stats.wins++;
      stats.points += 3;
    } else {
      stats.losses++;
    }
  }
  return stats;
}

// ── Match Queries ────────────────────────────────────────────────────

export function searchMatches(query: MatchQuery): NormalizedMatch[] {
  let matches = getMatches();
  const team = query.team ? normalizeTeam(query.team).key : null;
  const opponent = query.opponent ? normalizeTeam(query.opponent).key : null;

  if (team && opponent) {
    // Both specified: exactly those two teams
    matches = matches.filter(m =>
      (m.homeTeam === team && m.awayTeam === opponent) ||
      (m.homeTeam === opponent && m.awayTeam === team)
    );
  } else if (team) {
    matches = matches.filter(m => teamMatchesQuery(m, team));
  }

  if (query.competition) {
    const comp = query.competition.toLowerCase().replace(/[^a-z_]/g, '_');
    // Map common names
    const compKey = comp.includes('brasileir') || comp.includes('serie_a') ? 'brasileirao'
      : comp.includes('copa') && comp.includes('brasil') ? 'copa_do_brasil'
      : comp.includes('libertadores') ? 'libertadores'
      : comp;
    matches = matches.filter(m => m.competition === compKey);
  }

  if (query.season) {
    matches = matches.filter(m => m.season === query.season);
  }

  if (query.dateFrom) {
    matches = matches.filter(m => m.date >= query.dateFrom!);
  }
  if (query.dateTo) {
    matches = matches.filter(m => m.date <= query.dateTo!);
  }

  // Sort by date descending
  matches.sort((a, b) => b.date.localeCompare(a.date) || b.season - a.season);

  if (query.limit && query.limit > 0) {
    matches = matches.slice(0, query.limit);
  }

  return matches;
}

// ── Team Statistics ──────────────────────────────────────────────────

export function getTeamRecord(teamName: string): TeamRecord | null {
  const normalized = normalizeTeam(teamName);
  if (!normalized.key) return null;

  const allMatches = getMatches().filter(m => teamMatchesQuery(m, normalized.key));
  if (allMatches.length === 0) return null;

  const homeMatches = allMatches.filter(m => m.homeTeam === normalized.key);
  const awayMatches = allMatches.filter(m => m.awayTeam === normalized.key);

  // Per-competition stats
  const compMap = new Map<string, NormalizedMatch[]>();
  for (const m of allMatches) {
    const existing = compMap.get(m.competition) || [];
    existing.push(m);
    compMap.set(m.competition, existing);
  }
  const compStats: Record<string, TeamStats> = {};
  for (const [comp, ms] of compMap) {
    compStats[comp] = calcStats(normalized.key, normalized.display, ms);
  }

  return {
    ...calcStats(normalized.key, normalized.display, allMatches),
    homeStats: calcStats(normalized.key, normalized.display, homeMatches),
    awayStats: calcStats(normalized.key, normalized.display, awayMatches),
    competitions: compStats,
  };
}

// ── Head-to-Head ─────────────────────────────────────────────────────

export function getHeadToHead(team1: string, team2: string): HeadToHead | null {
  const t1 = normalizeTeam(team1);
  const t2 = normalizeTeam(team2);
  if (!t1.key || !t2.key) return null;

  const matches = getMatches().filter(m =>
    (m.homeTeam === t1.key && m.awayTeam === t2.key) ||
    (m.homeTeam === t2.key && m.awayTeam === t1.key)
  ).sort((a, b) => b.date.localeCompare(a.date));

  if (matches.length === 0) return null;

  let t1Wins = 0, t2Wins = 0, draws = 0, t1Goals = 0, t2Goals = 0;

  for (const m of matches) {
    if (m.homeTeam === t1.key) {
      t1Goals += m.homeGoal;
      t2Goals += m.awayGoal;
      if (m.homeGoal > m.awayGoal) t1Wins++;
      else if (m.homeGoal < m.awayGoal) t2Wins++;
      else draws++;
    } else {
      t1Goals += m.awayGoal;
      t2Goals += m.homeGoal;
      if (m.awayGoal > m.homeGoal) t1Wins++;
      else if (m.awayGoal < m.homeGoal) t2Wins++;
      else draws++;
    }
  }

  return {
    team1: t1.key,
    team1Display: t1.display,
    team2: t2.key,
    team2Display: t2.display,
    totalMatches: matches.length,
    team1Wins: t1Wins,
    team2Wins: t2Wins,
    draws,
    team1Goals: t1Goals,
    team2Goals: t2Goals,
    matches,
  };
}

// ── Player Queries ───────────────────────────────────────────────────

export function searchPlayers(query: PlayerQuery): NormalizedPlayer[] {
  let players = getPlayers();

  if (query.name) {
    const q = query.name.toLowerCase();
    players = players.filter(p => p.name.toLowerCase().includes(q));
  }

  if (query.nationality) {
    const nat = query.nationality.toLowerCase().trim();
    players = players.filter(p => p.nationality.toLowerCase() === nat);
  }

  if (query.club) {
    const club = normalizeTeam(query.club).key;
    if (club) {
      players = players.filter(p => p.club.includes(club));
    }
  }

  if (query.position) {
    const pos = query.position.toUpperCase().trim();
    players = players.filter(p => {
      const pPos = p.position.toUpperCase();
      // Allow partial match (e.g., "CDM" matches "CDM", "CB" matches "CB")
      return pPos.includes(pos);
    });
  }

  if (query.minRating !== undefined) {
    players = players.filter(p => p.overall >= query.minRating!);
  }
  if (query.maxRating !== undefined) {
    players = players.filter(p => p.overall <= query.maxRating!);
  }

  // Sort
  const sortBy = query.sortBy || '-overall';
  const desc = sortBy.startsWith('-');
  const field = desc ? sortBy.slice(1) : sortBy;

  players.sort((a, b) => {
    const aVal = (a as any)[field] ?? 0;
    const bVal = (b as any)[field] ?? 0;
    return desc ? bVal - aVal : aVal - bVal;
  });

  if (query.limit && query.limit > 0) {
    players = players.slice(0, query.limit);
  }

  return players;
}

// ── Competition Standings ────────────────────────────────────────────

const BRAZILIAN_POINTS_PER_WIN = 3;

export function getStandings(competition: string, season: number): StandingEntry[] {
  const compKey = competition.toLowerCase().replace(/[^a-z_]/g, '_');
  const cKey = compKey.includes('brasileir') || compKey.includes('serie_a') ? 'brasileirao'
    : compKey.includes('copa') && compKey.includes('brasil') ? 'copa_do_brasil'
    : compKey.includes('libertadores') ? 'libertadores'
    : compKey;

  const matches = getMatches().filter(m => m.competition === cKey && m.season === season);
  if (matches.length === 0) return [];

  // Accumulate per team
  const map = new Map<string, { display: string; played: number; wins: number; draws: number; losses: number; gf: number; ga: number }>();
  for (const m of matches) {
    for (const [team, display, gf, ga] of [
      [m.homeTeam, m.homeTeamDisplay, m.homeGoal, m.awayGoal] as const,
      [m.awayTeam, m.awayTeamDisplay, m.awayGoal, m.homeGoal] as const,
    ]) {
      let entry = map.get(team);
      if (!entry) {
        entry = { display, played: 0, wins: 0, draws: 0, losses: 0, gf: 0, ga: 0 };
        map.set(team, entry);
      }
      entry.played++;
      entry.gf += gf;
      entry.ga += ga;
      if (gf > ga) entry.wins++;
      else if (gf < ga) entry.losses++;
      else entry.draws++;
    }
  }

  // Convert to standings
  const standings: StandingEntry[] = [...map.entries()].map(([team, e]) => ({
    position: 0,
    team,
    teamDisplay: e.display,
    played: e.played,
    wins: e.wins,
    draws: e.draws,
    losses: e.losses,
    goalsFor: e.gf,
    goalsAgainst: e.ga,
    goalDifference: e.gf - e.ga,
    points: e.wins * BRAZILIAN_POINTS_PER_WIN + e.draws,
  }));

  // Sort: points DESC, wins DESC, goal diff DESC, goals for DESC
  standings.sort((a, b) =>
    b.points - a.points ||
    b.wins - a.wins ||
    b.goalDifference - a.goalDifference ||
    b.goalsFor - a.goalsFor
  );

  standings.forEach((s, i) => { s.position = i + 1; });

  return standings;
}

// ── Statistical Queries ──────────────────────────────────────────────

export function getBiggestWins(competition?: string, limit = 10): NormalizedMatch[] {
  let matches = getMatches();
  if (competition) {
    const cKey = competition.toLowerCase().replace(/[^a-z_]/g, '_');
    matches = matches.filter(m => m.competition === cKey || m.competition.includes(cKey));
  }
  // Only consider wins (not draws)
  matches = matches.filter(m => m.homeGoal !== m.awayGoal);
  matches.sort((a, b) => {
    const aDiff = Math.abs(a.homeGoal - a.awayGoal);
    const bDiff = Math.abs(b.homeGoal - b.awayGoal);
    return bDiff - aDiff || (b.homeGoal + b.awayGoal) - (a.homeGoal + a.awayGoal);
  });
  return matches.slice(0, limit);
}

export function getAverageGoals(competition?: string, season?: number): { avgGoalsPerMatch: number; totalMatches: number; totalGoals: number } {
  let matches = getMatches();
  if (competition) {
    const cKey = competition.toLowerCase().replace(/[^a-z_]/g, '_');
    matches = matches.filter(m => m.competition === cKey);
  }
  if (season) {
    matches = matches.filter(m => m.season === season);
  }
  const totalGoals = matches.reduce((sum, m) => sum + m.homeGoal + m.awayGoal, 0);
  return {
    avgGoalsPerMatch: matches.length > 0 ? Math.round((totalGoals / matches.length) * 100) / 100 : 0,
    totalMatches: matches.length,
    totalGoals,
  };
}

export function getHomeAwayStats(competition?: string, season?: number): {
  homeWins: number;
  awayWins: number;
  draws: number;
  totalMatches: number;
  homeWinRate: number;
} {
  let matches = getMatches();
  if (competition) {
    const cKey = competition.toLowerCase().replace(/[^a-z_]/g, '_');
    matches = matches.filter(m => m.competition === cKey);
  }
  if (season) {
    matches = matches.filter(m => m.season === season);
  }

  let homeWins = 0, awayWins = 0, draws = 0;
  for (const m of matches) {
    if (m.homeGoal > m.awayGoal) homeWins++;
    else if (m.awayGoal > m.homeGoal) awayWins++;
    else draws++;
  }

  return {
    homeWins,
    awayWins,
    draws,
    totalMatches: matches.length,
    homeWinRate: matches.length > 0 ? Math.round((homeWins / matches.length) * 1000) / 10 : 0,
  };
}

export function getTopScoringTeams(competition?: string, season?: number, limit = 10): { team: string; teamDisplay: string; goals: number; matches: number }[] {
  let matches = getMatches();
  if (competition) {
    const cKey = competition.toLowerCase().replace(/[^a-z_]/g, '_');
    matches = matches.filter(m => m.competition === cKey);
  }
  if (season) {
    matches = matches.filter(m => m.season === season);
  }

  const map = new Map<string, { display: string; goals: number; matches: number }>();
  for (const m of matches) {
    for (const [team, display, goals] of [
      [m.homeTeam, m.homeTeamDisplay, m.homeGoal] as const,
      [m.awayTeam, m.awayTeamDisplay, m.awayGoal] as const,
    ]) {
      let entry = map.get(team);
      if (!entry) {
        entry = { display, goals: 0, matches: 0 };
        map.set(team, entry);
      }
      entry.goals += goals;
      entry.matches++;
    }
  }

  return [...map.entries()]
    .map(([team, e]) => ({ team, teamDisplay: e.display, goals: e.goals, matches: e.matches }))
    .sort((a, b) => b.goals - a.goals)
    .slice(0, limit);
}

export function getTeamBestAwayRecord(limit = 10): { team: string; teamDisplay: string; awayWins: number; awayMatches: number; awayWinRate: number }[] {
  const matches = getMatches();
  const map = new Map<string, { display: string; wins: number; total: number }>();

  for (const m of matches) {
    let entry = map.get(m.awayTeam);
    if (!entry) {
      entry = { display: m.awayTeamDisplay, wins: 0, total: 0 };
      map.set(m.awayTeam, entry);
    }
    entry.total++;
    if (m.awayGoal > m.homeGoal) entry.wins++;
  }

  return [...map.entries()]
    .filter(([, e]) => e.total >= 5)
    .map(([team, e]) => ({
      team,
      teamDisplay: e.display,
      awayWins: e.wins,
      awayMatches: e.total,
      awayWinRate: Math.round((e.wins / e.total) * 1000) / 10,
    }))
    .sort((a, b) => b.awayWinRate - a.awayWinRate)
    .slice(0, limit);
}

export function getCompetitionList(): string[] {
  const comps = new Set<string>();
  for (const m of getMatches()) {
    comps.add(m.competition);
  }
  return [...comps].sort();
}

export function getSeasonList(competition?: string): number[] {
  let matches = getMatches();
  if (competition) {
    const cKey = competition.toLowerCase().replace(/[^a-z_]/g, '_');
    matches = matches.filter(m => m.competition === cKey);
  }
  const seasons = new Set<number>();
  for (const m of matches) {
    seasons.add(m.season);
  }
  return [...seasons].sort((a, b) => b - a);
}

File: experiments-local/experiment-17/runs/agent=omp_language=python_model=openrouter/moonshotai/kimi-k2.7-code_tooling=none/rep3/brazilian_soccer_mcp/queries.py
"""
Brazilian Soccer MCP - Query Engine.

This module implements the high-level query operations exposed by the MCP
server.  It is intentionally decoupled from ``server.py`` so that the same
logic can be unit-tested without spinning up an MCP transport.

All public functions accept raw user-facing strings and return either a
formatted markdown string or a serializable Python structure.  Team-name
normalization is performed internally.
"""

from __future__ import annotations

import json
from datetime import date, datetime
from typing import Any

import pandas as pd

from brazilian_soccer_mcp.data_store import DataStore, get_data_store
from brazilian_soccer_mcp.team_normalizer import normalize_team_name


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------
def _format_date(value: pd.Timestamp | Any) -> str:
    """Return a human-readable date string."""
    if pd.isna(value):
        return "Unknown"
    return pd.to_datetime(value).strftime("%Y-%m-%d")

def _match_line(row: pd.Series) -> str:
    """Format a single match as a markdown bullet."""
    competition = row.get("competition", "")
    round_label = row.get("round", "")
    stage = row.get("stage", "")
    extra = ""
    if competition:
        extra += f" ({competition}"
        if stage and str(stage).lower() not in ("nan", "none", ""):
            extra += f" {stage}"
        elif round_label and str(round_label).lower() not in ("nan", "none", ""):
            extra += f" Round {round_label}"
        extra += ")"
    hg = int(row["home_goal"]) if pd.notna(row["home_goal"]) else 0
    ag = int(row["away_goal"]) if pd.notna(row["away_goal"]) else 0
    return (
        f"- {_format_date(row['date'])}: "
        f"{row['home_team']} {hg}-{ag} "
        f"{row['away_team']}{extra}"
    )


def _filter_matches(
    store: DataStore,
    team: str | None = None,
    opponent: str | None = None,
    date_from: str | None = None,
    date_to: str | None = None,
    competition: str | None = None,
    season: int | None = None,
) -> pd.DataFrame:
    """Return a filtered copy of the matches DataFrame."""
    df = store.matches.copy()

    if team:
        canonical = normalize_team_name(team)
        df = df[(df["home_team"] == canonical) | (df["away_team"] == canonical)]

    if opponent:
        canonical_opp = normalize_team_name(opponent)
        df = df[
            (df["home_team"] == canonical_opp) | (df["away_team"] == canonical_opp)
        ]

    if competition:
        df = df[df["competition"].str.lower() == competition.lower()]

    if season is not None:
        df = df[df["season"] == season]

    if date_from:
        df = df[df["date"] >= pd.to_datetime(date_from)]
    if date_to:
        df = df[df["date"] <= pd.to_datetime(date_to)]

    return df


# ---------------------------------------------------------------------------
# Match queries
# ---------------------------------------------------------------------------
def find_matches(
    store: DataStore | None = None,
    team: str | None = None,
    opponent: str | None = None,
    date_from: str | None = None,
    date_to: str | None = None,
    competition: str | None = None,
    season: int | None = None,
    limit: int = 20,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Search matches by team, opponent, date range, competition and season."""
    store = store or get_data_store()
    df = _filter_matches(
        store,
        team=team,
        opponent=opponent,
        date_from=date_from,
        date_to=date_to,
        competition=competition,
        season=season,
    )

    total = len(df)
    df = df.head(limit)

    if response_format.lower() == "json":
        return {
            "total": total,
            "returned": len(df),
            "matches": df.replace({pd.NaT: None}).to_dict(orient="records"),
        }

    header = []
    if team and opponent:
        header.append(f"Matches between {normalize_team_name(team)} and {normalize_team_name(opponent)}")
    elif team:
        header.append(f"Matches for {normalize_team_name(team)}")
    elif opponent:
        header.append(f"Matches involving {normalize_team_name(opponent)}")
    else:
        header.append("Matches")

    if competition:
        header[-1] += f" ({competition})"
    if season:
        header[-1] += f" {season}"

    lines = [f"# {header[0]}", f"Showing {len(df)} of {total} matches", ""]
    for _, row in df.iterrows():
        lines.append(_match_line(row))
    return "\n".join(lines)


def get_head_to_head(
    store: DataStore | None = None,
    team_a: str = "",
    team_b: str = "",
    limit: int = 20,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Return the head-to-head history between two teams."""
    store = store or get_data_store()
    canonical_a = normalize_team_name(team_a)
    canonical_b = normalize_team_name(team_b)

    df = store.matches.copy()
    df = df[
        (
            (df["home_team"] == canonical_a) & (df["away_team"] == canonical_b)
        )
        | (
            (df["home_team"] == canonical_b) & (df["away_team"] == canonical_a)
        )
    ].sort_values("date", ascending=False)

    wins_a = draws = wins_b = 0
    for _, row in df.iterrows():
        if row["home_team"] == canonical_a:
            if row["home_goal"] > row["away_goal"]:
                wins_a += 1
            elif row["home_goal"] < row["away_goal"]:
                wins_b += 1
            else:
                draws += 1
        else:
            if row["home_goal"] > row["away_goal"]:
                wins_b += 1
            elif row["home_goal"] < row["away_goal"]:
                wins_a += 1
            else:
                draws += 1

    total = len(df)
    df = df.head(limit)

    if response_format.lower() == "json":
        return {
            "team_a": canonical_a,
            "team_b": canonical_b,
            "total_matches": total,
            f"{canonical_a}_wins": wins_a,
            "draws": draws,
            f"{canonical_b}_wins": wins_b,
            "matches": df.replace({pd.NaT: None}).to_dict(orient="records"),
        }

    lines = [
        f"# {canonical_a} vs {canonical_b}",
        f"Head-to-head in dataset: {canonical_a} {wins_a} wins, {canonical_b} {wins_b} wins, {draws} draws",
        f"Showing {len(df)} of {total} matches:",
        "",
    ]
    for _, row in df.iterrows():
        lines.append(_match_line(row))
    return "\n".join(lines)


# ---------------------------------------------------------------------------
# Team queries
# ---------------------------------------------------------------------------
def get_team_stats(
    store: DataStore | None = None,
    team: str = "",
    season: int | None = None,
    competition: str | None = None,
    venue: str | None = None,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Return wins, draws, losses and goals for a team."""
    store = store or get_data_store()
    canonical = normalize_team_name(team)
    stats = store.team_stats(canonical, season=season, competition=competition, venue=venue)

    if response_format.lower() == "json":
        return stats

    venue_label = venue.title() if venue else "Overall"
    season_label = str(season) if season else "all seasons"
    comp_label = competition if competition else "all competitions"
    return (
        f"# {canonical} {venue_label} Record ({comp_label}, {season_label})\n"
        f"- Matches: {stats['matches']}\n"
        f"- Wins: {stats['wins']}, Draws: {stats['draws']}, Losses: {stats['losses']}\n"
        f"- Goals For: {stats['goals_for']}, Goals Against: {stats['goals_against']}\n"
        f"- Goal Difference: {stats.get('goal_difference', 0)}\n"
        f"- Win rate: {stats['win_rate']}%"
    )


def list_teams(
    store: DataStore | None = None,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Return the list of teams present in the match data."""
    store = store or get_data_store()
    teams = sorted(
        set(store.matches["home_team"].dropna())
        | set(store.matches["away_team"].dropna())
    )
    teams = [t for t in teams if t]

    if response_format.lower() == "json":
        return {"teams": teams, "count": len(teams)}

    lines = [f"# Teams in dataset ({len(teams)})", ""]
    lines.extend(f"- {t}" for t in teams)
    return "\n".join(lines)


# ---------------------------------------------------------------------------
# Player queries
# ---------------------------------------------------------------------------
def search_players(
    store: DataStore | None = None,
    name: str | None = None,
    nationality: str | None = None,
    club: str | None = None,
    position: str | None = None,
    min_overall: int | None = None,
    limit: int = 20,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Search the FIFA player database."""
    store = store or get_data_store()
    df = store.players.copy()

    if name:
        df = df[df["name"].str.contains(name, case=False, na=False)]
    if nationality:
        df = df[
            df["nationality"]
            .str.strip()
            .str.lower()
            .eq(nationality.strip().lower())
        ]
    if club:
        canonical = normalize_team_name(club)
        df = df[
            (df["club"].str.lower() == canonical.lower())
            | df["club_raw"].str.contains(club, case=False, na=False)
        ]
    if position:
        df = df[df["position"].str.contains(position, case=False, na=False)]
    if min_overall is not None:
        df = df[df["overall"] >= min_overall]

    df = df.sort_values("overall", ascending=False).head(limit)

    if response_format.lower() == "json":
        return {
            "total_matching": len(df),
            "players": df.fillna("").to_dict(orient="records"),
        }

    lines = [f"# Player Search Results ({len(df)} found)", ""]
    for _, row in df.iterrows():
        club_display = row["club"] or row["club_raw"] or "Unknown"
        lines.append(
            f"- {row['name']} - Overall: {row['overall']}, "
            f"Position: {row['position']}, Club: {club_display}, "
            f"Nationality: {row['nationality']}"
        )
    return "\n".join(lines)


def top_brazilian_players(
    store: DataStore | None = None,
    limit: int = 10,
    at_brazilian_club: bool = False,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Return the highest-rated Brazilian players."""
    store = store or get_data_store()
    df = store.players[store.players["is_brazilian"]].copy()

    if at_brazilian_club:
        brazilian_clubs = set(store.matches["home_team"]) | set(store.matches["away_team"])
        df = df[df["club"].isin(brazilian_clubs)]

    df = df.sort_values("overall", ascending=False).head(limit)

    if response_format.lower() == "json":
        return {
            "count": len(df),
            "players": df.fillna("").to_dict(orient="records"),
        }

    lines = [f"# Top Brazilian Players{' at Brazilian Clubs' if at_brazilian_club else ''}", ""]
    for _, row in df.iterrows():
        club_display = row["club"] or row["club_raw"] or "Unknown"
        lines.append(
            f"- {row['name']} - Overall: {row['overall']}, "
            f"Position: {row['position']}, Club: {club_display}"
        )
    return "\n".join(lines)


# ---------------------------------------------------------------------------
# Competition queries
# ---------------------------------------------------------------------------
def get_standings(
    store: DataStore | None = None,
    season: int = 0,
    competition: str | None = None,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Return calculated league standings for a season."""
    store = store or get_data_store()
    df = store.standings(season, competition=competition)

    if response_format.lower() == "json":
        return {
            "season": season,
            "competition": competition,
            "standings": df.to_dict(orient="records"),
        }

    lines = [
        f"# {season} {competition or 'All Competitions'} Standings (calculated from matches)",
        "",
    ]
    for idx, row in df.iterrows():
        rank = int(idx) + 1
        marker = " - Champion" if rank == 1 else ""
        lines.append(
            f"{rank}. {row['team']} - {row['points']} pts "
            f"({row['wins']}W, {row['draws']}D, {row['losses']}L){marker}"
        )
    return "\n".join(lines)


def get_competition_winners(
    store: DataStore | None = None,
    season: int = 0,
    competition: str | None = None,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Return the team with the most points in a season/competition."""
    store = store or get_data_store()
    df = store.standings(season, competition=competition)

    if df.empty:
        result: dict[str, Any] = {
            "season": season,
            "competition": competition,
            "winner": None,
            "message": "No standings could be calculated for the requested season/competition.",
        }
        return result if response_format.lower() == "json" else result["message"]

    winner = df.iloc[0]
    result = {
        "season": season,
        "competition": competition,
        "winner": winner["team"],
        "points": int(winner["points"]),
        "record": f"{winner['wins']}W, {winner['draws']}D, {winner['losses']}L",
    }

    if response_format.lower() == "json":
        return result

    return (
        f"# {season} {competition or 'All Competitions'} Winner\n"
        f"{result['winner']} - {result['points']} pts ({result['record']})"
    )


# ---------------------------------------------------------------------------
# Statistical analysis queries
# ---------------------------------------------------------------------------
def get_biggest_wins(
    store: DataStore | None = None,
    competition: str | None = None,
    season: int | None = None,
    limit: int = 10,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Return the biggest victories in the dataset."""
    store = store or get_data_store()
    df = store.matches.copy()
    if competition:
        df = df[df["competition"].str.lower() == competition.lower()]
    if season is not None:
        df = df[df["season"] == season]

    df = df.dropna(subset=["home_goal", "away_goal"]).copy()
    df["margin"] = (df["home_goal"] - df["away_goal"]).abs()
    df = df.sort_values("margin", ascending=False).head(limit)

    if response_format.lower() == "json":
        return {
            "count": len(df),
            "matches": df.replace({pd.NaT: None}).to_dict(orient="records"),
        }

    lines = [f"# Biggest Victories{' (' + competition + ')' if competition else ''}", ""]
    for _, row in df.iterrows():
        lines.append(_match_line(row))
    return "\n".join(lines)


def get_average_goals(
    store: DataStore | None = None,
    competition: str | None = None,
    season: int | None = None,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Return average goals per match and home win rate."""
    store = store or get_data_store()
    df = store.matches.copy()
    if competition:
        df = df[df["competition"].str.lower() == competition.lower()]
    if season is not None:
        df = df[df["season"] == season]

    df = df.dropna(subset=["home_goal", "away_goal"])
    total_goals = df["home_goal"].sum() + df["away_goal"].sum()
    matches = len(df)
    avg = round(total_goals / matches, 2) if matches else 0.0

    home_wins = int((df["home_goal"] > df["away_goal"]).sum())
    home_win_rate = round(home_wins / matches * 100, 1) if matches else 0.0

    result = {
        "matches": matches,
        "average_goals_per_match": avg,
        "home_win_rate": home_win_rate,
    }
    if response_format.lower() == "json":
        return result

    return (
        f"# Goal Averages{' (' + competition + ')' if competition else ''}\n"
        f"- Matches analyzed: {matches}\n"
        f"- Average goals per match: {avg}\n"
        f"- Home win rate: {home_win_rate}%"
    )


def get_top_scorers(
    store: DataStore | None = None,
    season: int | None = None,
    limit: int = 10,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Return the teams with the most goals scored."""
    store = store or get_data_store()
    df = store.top_scorers(season=season, limit=limit)

    if response_format.lower() == "json":
        return {"count": len(df), "top_scorers": df.to_dict(orient="records")}

    season_label = str(season) if season else "all seasons"
    lines = [f"# Top Scoring Teams ({season_label})", ""]
    for idx, row in df.iterrows():
        lines.append(f"{int(idx) + 1}. {row['team']} - {int(row['goals'])} goals")
    return "\n".join(lines)


def compare_seasons(
    store: DataStore | None = None,
    season_a: int = 0,
    season_b: int = 0,
    competition: str | None = None,
    response_format: str = "markdown",
) -> str | dict[str, Any]:
    """Compare aggregate match statistics across two seasons."""
    store = store or get_data_store()

    def _season_stats(season: int) -> dict[str, Any]:
        df = store.matches[store.matches["season"] == season].copy()
        if competition:
            df = df[df["competition"].str.lower() == competition.lower()]
        df = df.dropna(subset=["home_goal", "away_goal"])
        matches = len(df)
        if matches == 0:
            return {"season": season, "matches": 0}
        home_wins = int((df["home_goal"] > df["away_goal"]).sum())
        draws = int((df["home_goal"] == df["away_goal"]).sum())
        away_wins = int((df["away_goal"] > df["home_goal"]).sum())
        return {
            "season": season,
            "matches": matches,
            "goals": int(df["home_goal"].sum() + df["away_goal"].sum()),
            "avg_goals": round(
                (df["home_goal"].sum() + df["away_goal"].sum()) / matches, 2
            ),
            "home_wins": home_wins,
            "draws": draws,
            "away_wins": away_wins,
            "home_win_rate": round(home_wins / matches * 100, 1),
        }

    stats_a = _season_stats(season_a)
    stats_b = _season_stats(season_b)
    result = {"season_a": stats_a, "season_b": stats_b}

    if response_format.lower() == "json":
        return result

    lines = [
        f"# Season Comparison: {season_a} vs {season_b}",
        "",
        f"| Metric | {season_a} | {season_b} |",
        "|---|---|---|",
    ]
    for key in ["matches", "goals", "avg_goals", "home_wins", "draws", "away_wins", "home_win_rate"]:
        label = key.replace("_", " ").title()
        lines.append(f"| {label} | {stats_a.get(key, 0)} | {stats_b.get(key, 0)} |")
    return "\n".join(lines)
